packages/marshal/src/encodeToSmallcaps.js

```diff
@@ -0,0 +1,495 @@
+      case 'string': {
+        if (!startsSpecial(encoding)) {
+          return encoding;
+        }
+        const c = encoding.charAt(0);
+        switch (c) {
+          case '!': {
+            // un-hilbert-ify the string
+            return encoding.slice(1);
+          }
+          case '%': {
+            return passableSymbolForName(encoding.slice(1));
+          }
+          case '#': {
+            switch (encoding) {
+              case '#undefined': {
+                return undefined;
+              }
+              case '#NaN': {
+                return NaN;
+              }
+              case '#Infinity': {
+                return Infinity;
+              }
+              case '#-Infinity': {
+                return -Infinity;
+              }
+              default: {
+                assert.fail(X`unknown constant "${q(encoding)}"`, TypeError);
+              }
+            }
+          }
+          case '+':
+          case '-': {
+            return BigInt(encoding);
+          }
+          case '$': {
+            const result = decodeRemotableFromSmallcaps(
+              encoding,
+              decodeFromSmallcaps,
+            );
+            if (passStyleOf(result) !== 'remotable') {
+              assert.fail(
+                X`internal: decodeRemotableFromSmallcaps option must return a remotable: ${result}`,
+              );
+            }
+            return result;
+          }
+          case '&': {
+            const result = decodePromiseFromSmallcaps(
+              encoding,
+              decodeFromSmallcaps,
+            );
+            if (passStyleOf(result) !== 'promise') {
+              assert.fail(
+                X`internal: decodePromiseFromSmallcaps option must return a promise: ${result}`,
+              );
+            }
+            return result;
+          }
+          default: {
+            assert.fail(
+              X`Special char ${q(c)} reserved for future use: ${encoding}`,
+            );
+          }
+        }
+      }
+      case 'object': {
+        if (encoding === null) {
+          return encoding;
+        }
+
+        if (isArray(encoding)) {
+          const result = [];
+          const { length } = encoding;
+          for (let i = 0; i < length; i += 1) {
+            result[i] = decodeFromSmallcaps(encoding[i]);
+          }
+          return result;
+        }
+
+        if (hasOwnPropertyOf(encoding, '#tag')) {
+          const { '#tag': tag, payload, ...rest } = encoding;
+          typeof tag === 'string' ||
+            assert.fail(
+              X`Value of "#tag", the tag, must be a string: ${encoding}`,
+            );
+          ownKeys(rest).length === 0 ||
+            assert.fail(
+              X`#tag record unexpected properties: ${q(ownKeys(rest))}`,
+            );
+          return makeTagged(tag, decodeFromSmallcaps(payload));
+        }
+
+        if (hasOwnPropertyOf(encoding, '#error')) {
+          const result = decodeErrorFromSmallcaps(
```

Comment:
Should there be an assertion that `encoding` has no extra keys, either here or in `decodeErrorFrom{Smallcaps,CapData}`, to mirror the `ownKeys(rest).length === 0 || assert.fail(…)` line for decoding tagged objects? Not having such a guard seems like a future hazard, although having it means more work if we ever do want to go beyond `name`, `message`, and `errorId` (and we don't currently have such a guard for either tagged objects or errors in decoding capdata).

Reply:
FWIW, I added the guard. But I am on the fence as well about whether it is the right decision.